### ISSUE TITLE
Markdown inside of hyperlink markdown doesn't work.

### ISSUE THREAD

**Issue by VertigoRay** [gitlabhq/gitlabhq#8643]
# Components

* GitLab 7.6.2
* GitLab Shell 2.4.0
* GitLab API v3
* Ruby 2.1.5p273
* Rails 4.1.1

# Description

When you use markdown inside of the markdown's square brackets to create bold, italics, etc ... it shows the html instead of generating the html. Shown in Example 1.

Additionally, when you use HTML inside of the markdown's square brackets to create bold, italics, etc ... it shows the no evidence of formatting changes. Shown in Example 2.

## Example 1

**Markdown:**

`the [fox *runs* fast](http://example.com).`

**Results:**

the [fox \<em\>runs\</em\> fast](http://example.com).

**Expected:**

the [fox *runs* fast](http://example.com).

## Example 2

**Markdown:**

`the [fox <em>runs</em> fast](http://example.com).`

**Results:**

the [fox runs fast](http://example.com).

**Expected:**

the [fox <em>runs</em> fast](http://example.com).

**Comment by rumpelsepp** [comment]
This seems more like a redcarpet issue.

**Comment by VertigoRay** [comment]
Are you saying it's *just* a cosmetic issue?

I didn't say this was mission critical, but it is an issue ...

**Comment by VertigoRay** [comment]
I submitted [the same issue over there](https://github.com/vmg/redcarpet/issues/456).

**Comment by rumpelsepp** [comment]
Yeah, it's the markdown library which is used to render the HTML. So maybe your issue is better placed in their bugtracker. :)

**Comment by robin850** [comment]
Hi there!

It looks like this is not a Redcarpet but an implementation issue in your custom render object. The problem is at [this line](https://github.com/gitlabhq/gitlabhq/blob/ac04665a8c35a9c96fe4fc6fe4391aa8ef703fd0/app/helpers/gitlab_markdown_helper.rb#L19) ; the `escape_once` method will escape the `<em>` tags.

I don't know what kind of behavior you are trying to simulate with this method but this is certainly not what you want ; this has been introduced by https://github.com/gitlabhq/gitlabhq/commit/496f88afe10f95a7aa64ea8ab10e57412f827283 but it looks like this was causing issues with images (see https://github.com/gitlabhq/gitlabhq/commit/f72dc7f7798f521d20a71fb465df57f0d66befb5).

**Comment by zxqfox** [comment]
One more example:

```
[`link`][]
[hello][`link`]

[`link`]: http://github.com
```

Expected
```
<a href="http://github.com"><code>link</code></a>
<a href="http://github.com">hello</a>
```

Actual
```
<a href="http://github.com">&lt;code&gt;link&lt;/code&gt;</a>
<a href="http://github.com">hello</a>
```

**Comment by Razer6** [comment]
Fixed in https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/619 Will be available in 7.11.

**Comment by zxqfox** [comment]
@Razer6 :+1: Thanks!
